Add germancloud support

r7vme opened the pull request:
There are multiple changes:
- added new configuration flag `--service.azure.cloud`
- added cloud support to azure client package (made it similar as done in Kubernetes)
- updated calico to 3.0.5 (as it was in changelog, but template was not updated)
- make `v2` active release and deprecate `v1`

Related: https://github.com/giantswarm/installations/pull/310
Related: https://github.com/giantswarm/k8scloudconfig/pull/368

Status: Tested in `tarantula` cluster creation started, but vmss fail to be created with "Internal server error" from Azure. Working on it.

r7vme commented on client/azure.go:
```diff
@@ -22,12 +22,21 @@ type AzureConfig struct {
+	subscriptionID          string
+	resourceManagerEndpoint string
+	servicePrincipalToken   *adal.ServicePrincipalToken
+}
```
Made this similar to https://github.com/kubernetes/kubernetes/blob/v1.10.2/pkg/cloudprovider/providers/azure/azure_client.go#L146-L155

Idea is to get oAuth token once (put to this object) and use it for clients.

r7vme commented on client/azure.go:
```diff
@@ -123,83 +150,81 @@ func ResponseWasNotFound(resp autorest.Response) bool {
-	client.Authorizer = autorest.NewBearerAuthorizer(spt)
+func newDeploymentsClient(config *azureClientConfig) (*resources.DeploymentsClient, error) {
+	client := resources.NewDeploymentsClientWithBaseURI(
+		config.resourceManagerEndpoint,
```
We need to use `config.resourceManagerEndpoint` of specific cloud. By default `NewDeploymentsClient` uses public one.

r7vme commented on service/controller/v1/resource/deployment/resource_test.go:
```diff
@@ -16,6 +16,7 @@ import (
 )
 
 var testAzure = setting.Azure{
+	Cloud: "AZUREPUBLICCLOUD",
```
Just a unit test fix for `v1`. Otherwise it does not pass validation.

r7vme commented on service/controller/v2/cloudconfig/template.go:
```diff
@@ -523,7 +523,7 @@ const (
 	cloudProviderConfFileOwner      = "root:root"
 	cloudProviderConfFilePermission = 0600
-	cloudProviderConfFileTemplate   = `cloud: AZUREPUBLICCLOUD
+	cloudProviderConfFileTemplate   = `cloud: {{ .Cloud }}
```
This is second important thing after Azure client. We need to put proper value in cloud provider.

xh3b4sd commented on client/azure.go:
```diff
@@ -123,83 +150,81 @@ func ResponseWasNotFound(resp autorest.Response) bool {
+		*oauthConfig,
+		config.ClientID,
+		config.ClientSecret,
+		env.ServiceManagementEndpoint)
```
Can we please not do this? It reads awfully IMHO.

r7vme replied:
done.

xh3b4sd replied:
You only fixed this in one place but the other places are still using the "undesired" style. So I would not like to wrap the function calls and I would also like to keep the error handling explicit. So instead of this. 

```
return foo(
  a, 
  b, 
  c, 
  d)
```

Lets please do this, **everywhere**. 

```
err := foo(a, b, c, d)
if err != nil {
  return microerror.Mask(err)
}

return nil
```

xh3b4sd commented on client/azure.go:
```diff
@@ -123,83 +150,81 @@ func ResponseWasNotFound(resp autorest.Response) bool {
+	} else {
+		env, err = azure.EnvironmentFromName(cloudName)
+	}
+	return &env, err
```
There should be error masking and I would question the need of returning a pointer here.

r7vme replied:
done. Returning value and masked error in this function and wrpa error above, where function called.

r7vme commented on service/controller/v2/version_bundle.go:
```diff
@@ -9,16 +9,21 @@ import (
+				Component:   "azure-operator",
+				Description: "Added German cloud support.",
+				Kind:        versionbundle.KindAdded,
+			},
```
done

r7vme commented:
@xh3b4sd @kopiczko addressed comments. Please take a look.

Changed back v2 to WIP, because we have to fix lack of MSI in german cloud before considering active release.

Want to merge this changes as they are all necessary.

xh3b4sd commented on client/azure.go:
```diff
@@ -123,83 +150,73 @@ func ResponseWasNotFound(resp autorest.Response) bool {
-	return adal.NewServicePrincipalToken(*oauthConfig, config.ClientID, config.ClientSecret, scope)
+	token, err := adal.NewServicePrincipalToken(*oauthConfig, config.ClientID, config.ClientSecret, env.ServiceManagementEndpoint)
+	if err != nil {
+		return nil, microerror.Maskf(err, "getting token")
```
Just FYI in practise the annotations are rather useless when masking existing errors. The masking mostly only makes sense when an error is created and a root cause is annotated. I noticed this is done that way in the azure-operator though. Just sharing my opinion here again.

r7vme commented:
CI ok, addressed all comments. Merging.
